### PR TITLE
Allow JSON notation `->` in column references for complex where conditions

### LINKED ISSUE #1863: Query JSON column in whereConditions
**Describe the bug**
If a database column is defined as JSON, it's currently not possible to query using whereConditions, since method `assertValidColumnReference` doesn't allow symbols `->` used as column name

**Expected behavior/Solution**
Having a table with a column of type JSON, it should be possible to define such column with JSON notation `->`

**Steps to reproduce**

1. Create a migration to add a table with a JSON column
2. Add related model and define a query to retrieve all models in that table using `@all` directive
3. Define an enum that has a value to query one JSON field (for example `enum ExampleColumn { VALUE @enum(value: "json_column->json_key") }`)
4. Add an argument to that query as `where: _ @whereConditions(columnsEnum: "ExampleColumn")`
5. Perform a query on that table after creating some mock data

**Additional info**
I think it should be possible to fix this changing the regex used with this one:  
`/^(?![0-9.-])([A-Za-z0-9_.-]|->)*$/`  
I'm opening related PR #1864 

**Lighthouse Version**
5.12.1

### PR DESCRIPTION
**Pull request by akyrey**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

Resolves #1863 

Add a test to query inner JSON. This test also required defining a new Scalar type JSON (ref. https://gist.github.com/lorado/0519972d6fcf01f1aa5179d09eb6a315)

**Changes**

Edited regex to check column name validity

**Breaking changes**

No breaking changes

### CODE REVIEW

File: src/WhereConditions/WhereConditionsBaseDirective.php [hunk 1]
```diff
@@ -134,7 +134,7 @@ public function manipulateArgDefinition(
         ObjectTypeDefinitionNode &$parentType
     ): void {
         if ($this->hasAllowedColumns()) {
-            $restrictedWhereConditionsName = ASTHelper::qualifiedArgType($argDefinition, $parentField, $parentType).$this->generatedInputSuffix();
+            $restrictedWhereConditionsName = ASTHelper::qualifiedArgType($argDefinition, $parentField, $parentType) . $this->generatedInputSuffix();
```
Comment (spawnia) [inline src/WhereConditions/WhereConditionsBaseDirective.php:137]: Please revert unnecessary formatting changes.

File: src/WhereConditions/WhereConditionsBaseDirective.php [hunk 2]
```diff
@@ -168,7 +168,7 @@ protected static function assertValidColumnReference(string $column): void
         // - must not start with a digit, dot or hyphen
         // - must contain only alphanumerics, digits, underscores, dots or hyphens
         // Dots are allowed to reference a column in a table: my_table.my_column.
-        $match = \Safe\preg_match('/^(?![0-9.-])[A-Za-z0-9_.-]*$/', $column);
+        $match = \Safe\preg_match('/^(?![0-9.-])([A-Za-z0-9_.-]|->)*$/', $column);
```
Comment (spawnia) [inline src/WhereConditions/WhereConditionsBaseDirective.php:171]: When would this be necessary? Please provide a test case that fails without this change and passes with it.

**Review by spawnia — CHANGES_REQUESTED** [review]

File: src/WhereConditions/WhereConditionsBaseDirective.php (hunk 2 above)
Comment (akyrey) [reply src/WhereConditions/WhereConditionsBaseDirective.php:171]: Is there a specific test model that I can modify adding a column or any is good?

File: src/WhereConditions/WhereConditionsBaseDirective.php (hunk 1 above)
Comment (akyrey) [reply src/WhereConditions/WhereConditionsBaseDirective.php]: Fixed

File: src/WhereConditions/WhereConditionsBaseDirective.php (hunk 2 above)
Comment (spawnia) [reply src/WhereConditions/WhereConditionsBaseDirective.php:171]: Any model with a string column should suffice, no?

Comment (akyrey) [reply src/WhereConditions/WhereConditionsBaseDirective.php:171]: Nope, would need a model with a JSON column in the database

Comment (spawnia) [reply src/WhereConditions/WhereConditionsBaseDirective.php:171]: Ok, just pick any one and add a nullable column.

**Comment by akyrey** [comment]
Should I update anything in the documentation?

**Comment by akyrey** [comment]
Sorry for multiple pushes for Style CI

**Comment by spawnia** [comment]
> Sorry for multiple pushes for Style CI

StyleCI has an option to download a patch.

**Comment by spawnia** [comment]
I saw you added a custom JSON type. It is not needed for purposes of this PR, but if you want to use one for your projects, I recommend https://github.com/mll-lab/graphql-php-scalars/blob/master/src/JSON.php
